WalletWasabi/Bases/PeriodicRunner.cs

```diff
@@ -1,132 +1,105 @@
 using Microsoft.Extensions.Hosting;
 using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
 using WalletWasabi.Logging;
 
 namespace WalletWasabi.Bases
 {
+	/// <summary>
+	/// <see cref="PeriodicRunner"/> is an extension of <see cref="BackgroundService"/> that is useful for tasks
+	/// that are supposed to repeat regularly.
+	/// </summary>
 	public abstract class PeriodicRunner : BackgroundService
 	{
+		private volatile TaskCompletionSource<bool>? _tcs;
+
 		protected PeriodicRunner(TimeSpan period)
 		{
-			TriggeringCts = new CancellationTokenSource();
-			TriggerLock = new object();
 			Period = period;
-			ResetLastException();
+			ExceptionTracker = new LastExceptionTracker();
 		}
 
-		private CancellationTokenSource TriggeringCts { get; set; }
-		private object TriggerLock { get; }
 		public TimeSpan Period { get; }
-		public Exception LastException { get; set; }
-		public long LastExceptionCount { get; set; }
-		public DateTimeOffset LastExceptionFirstAppeared { get; set; }
 
-		private void ResetLastException()
-		{
-			LastException = null;
-			LastExceptionCount = 0;
-			LastExceptionFirstAppeared = DateTimeOffset.MinValue;
-		}
+		private LastExceptionTracker ExceptionTracker { get; }
 
+		/// <summary>
+		/// Normally, <see cref="ActionAsync(CancellationToken)"/> user-action is called every time that <see cref="Period"/> elapses.
+		/// This method allows to expedite the process by interrupting the waiting process.
+		/// </summary>
+		/// <remarks>
+		/// If <see cref="ExecuteAsync(CancellationToken)"/> is not actually in waiting phase, this method call makes
+		/// sure that next waiting process will be omitted altogether.
+		/// </remarks>
 		public void TriggerRound()
 		{
-			lock (TriggerLock)
-			{
-				TriggeringCts?.Cancel();
-			}
+			// Note: All members of TaskCompletionSource<TResult> are thread-safe and may be used from multiple threads concurrently.
+			_tcs?.TrySetResult(true);
```

Comment:
It seems `_tcs` can be `null` only when the BackgroundService was not started because once it is initialized it is never null again. So, should we just ignore that fact we are calling `TriggerRound` for a non-running runner or should we fail with a "This runner is not running" exception?

Reply:
It's a tough call. I would vote for ignoring. I'll add a remark about that to the documentation.